rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNode.java

```diff
@@ -0,0 +1,172 @@
+        nodeDescription = "Produces incoming messages using GPS based geofencing",
+        nodeDetails = "Extracts latitude and longitude parameters from incoming message and returns different events " +
+                "based on configuration parameters")
+public class TbGpsMultiGeofencingActionNode extends AbstractGeofencingNode<TbGpsMultiGeofencingActionNodeConfiguration> {
+
+    private GeofencingProcessor geofencingProcessor;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        super.init(ctx, configuration);
+        geofencingProcessor = new GeofencingProcessor(ctx, this.config);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        List<EntityId> matchedGeofenceIds = findMatchedZones(ctx, msg);
+        GeofenceResponse geofenceResponse = geofencingProcessor.process(msg.getMetaDataTs(), ctx.getSelfId(), msg.getOriginator(), matchedGeofenceIds);
+        processGeofenceResponse(ctx, msg, msg.getOriginator(), geofenceResponse);
+        ctx.tellSuccess(msg);
+    }
+
+    private void processGeofenceResponse(TbContext ctx, TbMsg originalMsg, EntityId originatorId, GeofenceResponse geofenceResponse) {
+        TbMsgMetaData metaData = originalMsg.getMetaData().copy();
+        metaData.putValue("originatorId", originatorId.toString());
+        metaData.putValue("originatorName", getOriginatorName(ctx, originatorId));
+
+        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getEnteredGeofences(), "Entered");
+        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getLeftGeofences(), "Left");
+        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getInsideGeofences(), "Inside");
+        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getOutsideGeofences(), "Outside");
+    }
+
+    private void processGeofenceEvents(TbContext ctx, TbMsg originalMsg, TbMsgMetaData metaData, List<EntityId> geofences, String relationType) {
+        if (CollectionUtils.isEmpty(geofences)) {
+            return;
+        }
+        for (EntityId geofence : geofences) {
+            TbMsg tbMsg = TbMsg.newMsg(TbMsgType.valueOf(originalMsg.getType()), geofence, metaData, originalMsg.getData());
+            ctx.tellNext(tbMsg, relationType);
+        }
+    }
+
+    private String getOriginatorName(TbContext tbContext, EntityId originatorId) {
+        if (EntityType.ASSET.equals(originatorId.getEntityType())) {
+            return tbContext.getAssetService().findAssetById(tbContext.getTenantId(), (AssetId) originatorId).getName();
+        } else if (EntityType.DEVICE.equals(originatorId.getEntityType())) {
+            return tbContext.getDeviceService().findDeviceById(tbContext.getTenantId(), (DeviceId) originatorId).getName();
+        }
+        return null;
+    }
```

Comment:
1. This logic looks unrelated to the rule node goal of producing geofence events. This logic can be done with existing "originator fields" enrichment node.
2. findAssetById and findDeviceById blocking operations.

Reply:
I agree with you that the logic for retrieving name is unrelated to this node. I have removed it